### DISCUSSION TITLE
[Error: An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance which may provide additional details about the nature of the error.] {   digest: '1930860604' }

### DISCUSSION THREAD

**Discussion by Macfros** [vercel/next.js#57699] [Help]
### Summary

Iam trying to deploy my application to nextjs but I am getting this error. I am also getting a graphql error. but I dont understand this server component error. please help me solve this. 

the code for the main page is :- 
`import LoadMore from "@/components/LoadMore";
import { fetchAllProjects } from "@/lib/actions";
import Categories from "@/components/Categories";
import { ProjectInterface } from "@/common.types";
import ProjectCard from "@/components/ProjectCard";

type ProjectSearch = {
  projectSearch: {
    edges: { node: ProjectInterface }[];
    pageInfo: {
      hasPreviousPage: boolean;
      hasNextPage: boolean;
      startCursor: string;
      endCursor: string;
    };
  },
}

type SearchParams = {
  category?: string;
  endcursor?: string;
}

type Props = {
  searchParams: SearchParams;
}

export const dynamic = 'force-dynamic'; // force dynamic for this page so that it is always regenerated when a new project is created
export const dynamicParams = true;
export const revalidate = 0;

const Home = async ({ searchParams: { category, endcursor } }: Props) => {
  const data = await fetchAllProjects(category, endcursor) as ProjectSearch;

  const projectsToDisplay = data?.projectSearch?.edges || [];

  if(projectsToDisplay.length === 0) {
    return (
      <section className="flexStart flex-col paddings">
        <Categories />

        <p className="no-result-text text-center text-2xl">No projects found! 😵‍</p>
      </section>
    )
  }

  const pagination = data?.projectSearch?.pageInfo;

  return (
    <section className="flex-start flex-col paddings mb-16">
        <Categories />

        <section className="projects-grid">
          {projectsToDisplay.map(({ node }: { node: ProjectInterface }) => (
            <ProjectCard
              key={node?.id}
              id={node?.id}
              image={node?.image}
              title={node?.title}
              name={node?.createdBy?.name}
              avatarUrl={node?.createdBy?.avatarUrl}
              userId={node?.createdBy?.id}
            />
          ))}
        </section>

        <LoadMore
          startCursor={pagination.startCursor}
          endCursor={pagination.endCursor}
          hasPreviousPage={pagination.hasPreviousPage}
          hasNextPage={pagination.hasNextPage}
        />
    </section>
  )
}

export default Home
`

This is the code in the lib/actions

`export const fetchAllProjects = (category?: string, endcursor?: string) => {
  client.setHeader('x-api-key', apiKey);

  const query = category ? projectsQuery : getAllProjectsQuery;
  const variables = category ? { category, endcursor } : { endcursor };

  return makeGraphQLRequest(query, variables);
};
`

Can someone tell me in detail how to solve this problem. I am a new to this so detailed explanation and steps are appreciated. Thanks

### Example

https://github.com/Macfros/craftedVisions

**Comment by icyJoseph** [comment]
I think you need to add some more monitoring to your app. In the form of logs!

```ts
const makeGraphQLRequest = async (query: string, variables = {}) => {
  try {
    return await client.request(query, variables);
  } catch (err) {
    throw err;
  }
};
```

The catch branch should log the reason why the above did throw. Right now you are operating in the dark.

**Comment by justinbalaguer** [comment]
did you solve this?

**Comment by mailok** [comment]
I encountered this problem, but only in production. In my case, I'm throwing errors from Server Actions, and after almost completing the app, I found out that in development mode you can see the messages of your custom errors. However, in production, the error thrown from the server actions is changed to:

_"An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance which may provide additional details about the nature of the error."_

That's really bad. As I understand, it's for security reasons, because Next.js can't distinguish between errors intentionally thrown by the user and unexpected ones that may contain sensitive information not meant to reach the client.

**Reply by fitimbytyqi** [reply]
+1 , an ugly workaround I had to do with tanstack/query is returning the data from an external api and only throw the errors if there's any from the global onSuccess and catch it from global onError.

**Reply by thecodedlinda** [reply]
can u show an example of what this looks like?I am facing this issue in production as well

**Reply by fitimbytyqi** [reply]
@lordlinda I ended up using next-safe-actions or zsa, they solve this particular issue at its finest :)

**Reply by developer-vm-sixteen** [reply]
I am also facing this issue. How can I fix it?

**Reply by thecodedlinda** [reply]
@neutrix-developer-sixteen To handle errors thrown intentionally by the user, I decided not to handle them in the server component but in another component, so I just return them and throw them in another component.
Check out this article
https://joulev.dev/blogs/throwing-expected-errors-in-react-server-actions

**Reply by nghiepdev** [reply]
If you are invoking a Server Action imperatively, we can wrap the `startTransition` in a `try/catch` block, and handle the error from there.
```tsx
    <button
      disabled={pending}
      onClick={() => {
        startTransition(async () => {
          try {
            await serverActionWithError()
          } catch (e) {
            console.error(e);
          }
        });
      }}
    >
      Click Button
    </button>
```
